consensus/istanbul/backend/engine.go

```diff
@@ -272,29 +270,103 @@ func (sb *Backend) verifyCommittedSeals(chain consensus.ChainReader, header *typ
 		return errEmptyCommittedSeals
 	}
 
+	// Check the signatures on the current header
+	snap, err := sb.snapshot(chain, header.Number.Uint64()-1, header.ParentHash, parents)
+	if err != nil {
+		return err
+	}
 	validators := snap.ValSet.Copy()
-	// Check whether the committed seals are generated by parent's validators
+	err = sb.checkValidatorSignatures(header.Hash(), validators, extra.Bitmap, extra.CommittedSeal)
+	if err != nil {
+		return err
+	}
+
+	// Check parent signatures only after block 1.
+	// The genesis block is skipped since it has no parents
+	// The first block is also skipped since its parent
+	// is the genesis block which contains no parent signatures
+	if number > 1 {
+		sb.logger.Debug("verifyCommittedSeals: verifying parent block for block", "num", number)
+		var parentValidators istanbul.ValidatorSet
+		// parents of epoch blocks have different validator sets since they
+		// belong to the previous epoch.
+		if number%sb.config.Epoch == 0 {
+			sb.logger.Debug("verifyCommittedSeals: it's an epoch block!")
+			var parent *types.Header
+			var cache []*types.Header
+			if len(parents) > 1 {
+				cache = parents[:len(parents)-1]
+				parent = parents[len(parents)-1]
+				sb.logger.Debug("verifyCommittedSeals: was given a cache", "cache-len", len(cache), "parent-num-from-cache", parent.Number.Uint64())
+			} else {
+				cache = parents
+				parent = chain.GetHeaderByNumber(number - sb.config.Epoch)
+				// When in ultralight mode, sometimes we are unable to fetch the parent.
+				// In that case, we cannot fetch the validator set for the previous epoch,
+				// so we will assume it's valid.
+				if parent == nil {
+					sb.logger.Debug("verifyCommittedSeals: got nil parent", "num", number-sb.config.Epoch)
+					return nil
+				}
+				sb.logger.Debug("verifyCommittedSeals: no cache provided", "cache-len", len(cache), "parent-num", parent.Number.Uint64())
+			}
 
-	proposalSeal := istanbulCore.PrepareCommittedSeal(header.Hash())
-	// 1. Get committed seals from current header
-	myValidatorIndex, myValidator := validators.GetByAddress(sb.Address())
+			// Get the validator set from the parent block's snapshot.
+			// If in ultralight mode, the parent block is going to be the
```

Comment:
Actually, the way `snapshot` is written, it can handle arbitrary block numbers, even in ultralight mode. It does the logic of fetching only epoch blocks under the hood. So we should always just call snapshot here on the block before the parent block